### ISSUE TITLE
Allow XMPP Chat Server for Mastodon

### ISSUE THREAD

**Issue by shawneric** [tootsuite/mastodon#9181]
XMPP Chat provides a real-time chat solution for users who may wish to have an actual IM session with friends. This would be more streamlined and instant than DMs or Toots.

I have recently setup a working XMPP chat service with my diaspora* server and I'm very very interested to see if it'd be work as integrated with Mastodon as well.

**Comment by alxlg** [comment]
Do you mean DMs supporting XMPP like Facebook and Google Talk used to do?

**Comment by kvsjxd** [comment]
This can be useful

**Comment by vibhi19** [comment]
Gled fork has outh in place and you can login in with same username and password as your mastodon account but its not perfect for chat as you have to open a new link and then login to chat most user dont even use it.It would be better something can be added in webui and you can chat from there.

Although another issue i see is with mastodon is version fragmentation if mastodon get this feature how you gonna dm user on other server that are on older version some still use too old version

**Comment by vibhi19** [comment]
> Do you mean DMs supporting XMPP like Facebook and Google Talk used to do?

I think he mean that as of now dm are not encrypted

**Comment by alxlg** [comment]
@vibhi19 I don't think so, XMPP is a protocol to federate instant messaging platforms that happens to have an extension for end-to-end encryption called OMEMO.

Mastodon direct messages being compatible with XMPP means one can connect to his Mastodon account with a XMPP client and manage private conversations from there.

**Comment by trwnh** [comment]
@alxlg Adding XMPP to Mastodon would be a separate feature from the existing "direct" addressing option.

IMO there's no real reason for Mastodon to package its own XMPP server on top of the already included servers. What can be done instead is to make integration with an existing XMPP server much easier, e.g. with Prosody or ejabberd or Openfire.

In that regard, there are probably various modules for, say:
- using LDAP, which would allow provisioning accounts that can access both Mastodon and XMPP, at the cost of requiring registrations through the LDAP server instead of through Mastodon,
- or it could be possible to OAuth into XMPP with Mastodon credentials, e.g. by writing a module something like this one for Prosody: https://modules.prosody.im/mod_auth_oauthbearer.html

possibly tangentially related: #3165 #4800 #7958 could provide additional interoperability with other identity provider schemes

**Comment by vibhi19** [comment]
> @vibhi19 I don't think so, XMPP is a protocol to federate instant messaging platforms that happens to have an extension for end-to-end encryption called OMEMO.
> 
> Mastodon direct messages being compatible with XMPP means one can connect to his Mastodon account with a XMPP client and manage private conversations from there.

hmm that could work

**Comment by vibhi19** [comment]
> @alxlg Adding XMPP to Mastodon would be a separate feature from the existing "direct" addressing option.
> 

Yes it is different although take inspiration from misskey. misskey has a feature called "Talk" something like that can be implemented.You can talk to user one on one like twitter has and still use  "direct" addressing and you can use  "direct" addressing as you use on mastodon

**Comment by Neustradamus** [comment]
Any news?

**Comment by rkingett** [comment]
What about making it work with Matrix instead?

**Comment by rokejulianlockhart** [comment]
@rkingett, https://github.com/mastodon/mastodon/issues/14576

**Comment by sjehuda** [comment]
Implementing an additional layer for XMPP - and consequently allowing to enable ActivityPub and XMPP or only one of these two - would allow many features that [XEP-0060: Publish-Subscribe](https://xmpp.org/extensions/xep-0060.html) offers; One of which might be to regulate your publishing destinations (See also "Syndication Feed Opt-out" at #22172).

Blasta, Jappix Me, Libervia, and Movim do it well.
I suggest that Mastodon would be a part of XMPP as well.

**Comment by sjehuda** [comment]
> What about making it work with Natrix instead?

1) This is not the same.
2) It does not allow to store contents.
3) It is related to intelligence agencies.
4) It is not a standard.
5) It discloses all metadata without the ability to opt-out.
6) It is of a low quality.
7) It is bad.
8) No intelligence agency which honors itself makes use of it.
9) It encourages criminal activities by people who generally have no criminal intentions.

### No offence, but...

By participating in a system which discloses any metadata, and having an abusive centralized "authority", so called, people who are using it appear to me like people who publicly go outside without underwear, so it would be possible to see the body textures, or with as less as clothing articles as possible in order to reveal themselves to others as if they were a public property.

Here is an issue amonst many other issues [Matrix - A Pit of Abuse with Government Ties](https://alltube.private.coffee/info?url=https://www.youtube.com/watch?v=W8KEuAEYjQ4).

---

### XMPP (Jabber)

#### Tried and tested. Independent. Privacy-focused.

1) XMPP is a standard.
2) XMPP is private.
3) XMPP does not discloses your metadata to anyone.
4) XMPP has a mature authorization mechanism.
5) XMPP is lightweight on system resources.
6) **XMPP has a mature publishing platform.**
7) **XMPP allows to attach Atom Syndication Format feeds and consequently rich text HTML documents natively.**
* [XEP-0277: Microblogging over XMPP](https://xmpp.org/extensions/xep-0277.html)
* [XEP-0472: PubSub Social Feed](https://xmpp.org/extensions/xep-0472.html)
8) XMPP is not related to spy agencies.
9) XMPP is used by agencies, corporations, and people who respect their privacy and cause.

As I have wrote at https://github.com/mastodon/mastodon/issues/20924#issuecomment-2320032750

> I am a criminal and corporate lawyer, and amongst my clients XMPP is the "gold standard", so to speak, of communication.

> If someone does not use XMPP, then he is probably not my client, nor someone who my clients would be interested in engaging on business with.

I also leave only 900 meters from the spy agency which is related to the so called platform that you have suggested, and ever since its creation, that spy agency appears to be as a joke to me and to my associates.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/mastodon/mastodon/issues/9181#issuecomment-2320041946

@sjehuda, are you proposing an XMPP backend for regular post subscription, as an alternative to ActivityPub? If so, I suggest that you create a new issue, since that's rather substantially different to the current DM functionality, don't you think?

**Comment by Leatherface75** [comment]
I think Matrix support would be better because i think most people use that nowadays.

**Comment by sjehuda** [comment]
Leatherface75.
Your comment is off topic.
And I also do not agree with your statement.
Please hide your comment.

**Comment by CitizenPrayer** [comment]
Six years and the notion of end to end encrypted communications in DMs is now the industry standard. Why hasn't it been setup yet?

**Comment by RokeJulianLockhart** [comment]
> https://github.com/mastodon/mastodon/issues/9181#issuecomment-2354271964

@CitizenPrayer, because Mastodon isn't designed as a DM platform, and the current implementation uses the same logic as regular posting, whereas encrypted, XMPP-compatible chats would necessitate a significant overhaul.

I suggest you become satisfied with Pidgin with XMPP, and Element with Matrix for the foreseeable future.

**Comment by SamWhited** [comment]
Way *way* off topic, but I don't know why people keep spreading that you should use Pidgin. The XMPP implementation in Pidgin hasn't gotten any significant updates in years and is stuck in the early 2000's from a protocol/features level. Pretty sure it single handedly has created the myth that XMPP doesn't have any modern features. Use something that's actually maintained :) (Dino or Gajim come to mind, but I'm only on Linux there are other things for other platforms and your tastes may vary).

**Comment by RokeJulianLockhart** [comment]
> https://github.com/mastodon/mastodon/issues/9181#issuecomment-2361107840

@SamWhited, I didn't know of anything else. Thanks for the suggestions.

**Comment by trwnh** [comment]
> a significant benefit of building XMPP into Mastodon serverside is letting you keep the same handle

You don't need to build XMPP into Mastodon to accomplish this. You can already link both Mastodon and your XMPP server to the same external identity, through SSO/LDAP/similar. (This isn't to say that there aren't any benefits to adding XMPP support, but "same handle" is not one of them.)
